migrator/migrations/m_181_to_m_182_remove_default_scope_manager_role/migration_impl.go

```diff
@@ -0,0 +1,327 @@
+
+	apiTokenStorage := apiTokenStore.New(database.PostgresDB)
+	groupStorage := groupStore.New(database.PostgresDB)
+	permissionSetStorage := permissionSetStore.New(database.PostgresDB)
+	roleStorage := roleStore.New(database.PostgresDB)
+
+	var err *multierror.Error
+	canRemoveRole := true
+	canRemovePermissionSet := true
+	// Check whether the old default role is used
+	usesScopeManagerRole, useRoleCheckErr := isScopeManagerRoleReferenced(ctx, apiTokenStorage, groupStorage)
+	if useRoleCheckErr != nil {
+		canRemoveRole = false
+		err = multierror.Append(err, useRoleCheckErr)
+	}
+	if usesScopeManagerRole {
+		addRoleErr := addDeprecatedScopeManagerRole(ctx, roleStorage)
+		if addRoleErr != nil {
+			canRemoveRole = false
+			err = multierror.Append(err, addRoleErr)
+		} else {
+			tokenUpdateErr := updateScopeManagerRoleReferencesInAPITokens(ctx, apiTokenStorage)
+			if tokenUpdateErr != nil {
+				canRemoveRole = false
+				err = multierror.Append(err, tokenUpdateErr)
+			}
+			groupUpdateErr := updateScopeManagerRoleReferencesInGroups(ctx, groupStorage)
+			if groupUpdateErr != nil {
+				canRemoveRole = false
+				err = multierror.Append(err, tokenUpdateErr)
+			}
+		}
+	}
+	if canRemoveRole {
+		roleRemoveErr := removeDefaultScopeManagerRole(ctx, roleStorage)
+		if roleRemoveErr != nil {
+			canRemovePermissionSet = false
+			err = multierror.Append(err, roleRemoveErr)
+		}
+	} else {
+		canRemovePermissionSet = false
+	}
+	// Check for permission set references
+	usesScopeManagerPermissionSet, usePermissionSetCheckErr := isScopeManagerPermissionSetReferenced(ctx, roleStorage)
+	if usePermissionSetCheckErr != nil {
+		canRemovePermissionSet = false
+		err = multierror.Append(err, usePermissionSetCheckErr)
+	}
+	if usesScopeManagerPermissionSet {
+		// The Role -> PermissionSet reference is using the PermissionSet ID which will be left untouched.
+		// Therefore, there is no need to update the PermissionSet reference in the roles.
+		// The permission set Name and Description will be updated.
+		updatePermissionSetErr := updateScopeManagerPermissionSet(ctx, permissionSetStorage)
+		if updatePermissionSetErr != nil {
+			err = multierror.Append(err, updatePermissionSetErr)
+		}
+	} else {
+		if canRemovePermissionSet {
+			deletePermissionSetErr := removeDefaultScopeManagerPermissionSet(ctx, permissionSetStorage)
+			if deletePermissionSetErr != nil {
+				err = multierror.Append(err, deletePermissionSetErr)
+			}
+		}
+	}
+	return err.ErrorOrNil()
+}
+
+func apiTokenHasScopeManagerRole(obj *storage.TokenMetadata) bool {
+	tokenRoles := obj.GetRoles()
+	for _, roleName := range tokenRoles {
+		if roleName == ScopeManagerRoleName {
+			return true
+		}
+	}
+	return false
+}
+
+func groupHasScopeManagerRole(obj *storage.Group) bool {
+	return obj.GetRoleName() == ScopeManagerRoleName
+}
+
+func isScopeManagerRoleReferenced(ctx context.Context, apiTokenStorage apiTokenStore.Store, groupStorage groupStore.Store) (bool, error) {
```

Comment:
Instead of only returning a bool that references are there, since we anyways walk through the stores here, why not return the IDs of objects referencing the roles, and during update / delete step we only attempt to change those?

Reply:
Good idea :-).